2082

edvbld opened the pull request:
Hi all,

please review this patch that ensures that we pass the correct `.jcheck/conf` to [AdditionalConfiguration.java](https://github.com/openjdk/skara/blob/master/bots/pr/src/main/java/org/openjdk/skara/bots/pr/AdditionalConfiguration.java). The `.jcheck/conf` should either come from the "overriding" `.jcheck/conf` (if the PR bot is configured to use that) or the pull request's target branch.

I have added that test that passes with this patch but fails without it.

Thanks,
Erik

erikj79 commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/AdditionalConfiguration.java:
```diff
@@ -31,14 +31,14 @@
 
 public class AdditionalConfiguration {
-    static List<String> get(ReadOnlyRepository repository, Hash hash, HostUser botUser, List<Comment> comments, boolean reviewMerge) throws IOException {
+    static List<String> get(Optional<JCheckConfiguration> conf, HostUser botUser, List<Comment> comments, boolean reviewMerge) throws IOException {
```
Using Optional as parameter isn't recommended, especially since we throw if it's empty. Is it truly optional here? I would think we should just `.orElseThrow()` at the parse call.

edvbld replied:
Thanks, removed the use of `Optional` in the latest version.

erikj79 commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:
```diff
@@ -1265,9 +1267,10 @@ private void checkStatus() {
                         pr.repository().forge().currentUser(), comments, reviewMerge);
-                checkablePullRequest.executeChecks(localHash, censusInstance, visitor, additionalConfiguration, checkablePullRequest.targetHash());
+                checkablePullRequest.executeChecks(localHash, censusInstance, visitor, additionalConfiguration, jcheckConfHash);
                 // Don't need to run the second round if confOverride is set.
```
Strictly speaking, should we regenerate the AdditionalConfiguration for the second run where we use the merged .jcheck/conf?

zhaosongzs replied:
Oh, right, I overlooked it when I was implementing the second run of jcheck feature.

edvbld replied:
This has been re-worked in the latest version

erikj79 commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:
```diff
@@ -1254,7 +1254,9 @@ private void checkStatus() {
             }
-            PullRequestCheckIssueVisitor visitor = checkablePullRequest.createVisitor(checkablePullRequest.targetHash());
+
+            Hash jcheckConfHash = checkablePullRequest.targetHash();
+            PullRequestCheckIssueVisitor visitor = checkablePullRequest.createVisitor(jcheckConfHash);
```
Any particular reason to spell out the types here?

edvbld replied:
Nope, thanks for catching

edvbld commented:
Ok, so I spent quite a bit of time going through the places where we parse `.jcheck/conf`. I figured out the following:

- for merge PRs we want to always test each commit in the PR with _both_ the target branch's `.jcheck/conf` and, if any commit in the PR updates `.jcheck/conf`, the `.jcheck/conf` present in the commit. This could be optimized to only test with the `.jcheck/conf` in the commit if the current commit, or any parents of the current commit, updated `.jcheck/conf`. But that seemed like a premature optimization given that the number of merge PR that changes `.jcheck/conf` most likely will be very few.
- for regular pull requests we want to ensure that we are always using the `.jcheck/conf` from the target branch _with_ any additional configuration added (or using an overriding `.jcheck/conf` is that is set up). To enable this I moved this logic into `CheckablePullRequest.parseJCheckConfiguration`.

The remainder of the patch is dealing with fall-out from the above two changes and aligning the code with them.

Thanks,
Erik

PS. The tests failed due to JBS being down

erikj79 commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckablePullRequest.java:
```diff
@@ -202,25 +204,30 @@ Hash amendManualReviewers(Hash commit, Namespace namespace, Hash original) throw
-        }
-        if (conf.isEmpty()) {
-            throw new RuntimeException("Failed to parse jcheck configuration at: " + hash + " with extra: " + additionalConfiguration);
+    Optional<JCheckConfiguration> parseJCheckConfiguration(Hash hash) throws IOException {
```
Every caller of this method is throwing IllegalStateException if jcheck conf isn't found. Did you consider just throwing here and not return Optional? I realize there is a bit more information in the exception message, but on the other hand, whoever reads it will have a stacktrace anyway.

erikj79 commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckablePullRequest.java:
```diff
@@ -202,25 +204,30 @@ Hash amendManualReviewers(Hash commit, Namespace namespace, Hash original) throw
+            JCheck.parseConfiguration(confOverride, additional);
+    }
+
+    void executeChecks(Hash localHash, CensusInstance censusInstance, PullRequestCheckIssueVisitor visitor, JCheckConfiguration conf) throws IOException {
```
The parameter name `localHash` threw me off for a bit, maybe just `hash` in this context, unless I'm missing some context?

edvbld replied:
Sure, renamed it t to `hash` in the latest version

erikj79 reviewed (commented):
Overall I really like this change. It makes dealing with jcheck configs a lot clearer.

erikj79 commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:
```diff
@@ -1209,25 +1208,40 @@ private void checkStatus() {
+                        var targetVisitor = checkablePullRequest.createVisitor(targetJCheckConf, hash);
+                        checkablePullRequest.executeChecks(hash, censusInstance, targetVisitor, targetJCheckConf);
+
+                        var commitJCheckConf = checkablePullRequest.parseJCheckConfiguration(hash)
+                                                .orElseThrow(() -> new IllegalStateException("No .jcheck/conf present in tree for commit " + hash));
+                        var commitVisitor = checkablePullRequest.createVisitor(commitJCheckConf, hash);
```
Maybe we shouldn't concat the streams and instead differentiate the message between the different jcheck runs and include information about which jcheck conf was used.

zhaosongzs replied:
I agree with Erik here, it's better to include the jcheck conf information

zhaosongzs reviewed (commented):
This patch looks pretty good, just some minor issues to address.

edvbld commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:
```diff
@@ -1254,8 +1268,9 @@ private void checkStatus() {
-            PullRequestCheckIssueVisitor visitor = checkablePullRequest.createVisitor(checkablePullRequest.targetHash());
-            boolean needUpdateAdditionalProgresses = false;
+
+            var visitor = checkablePullRequest.createVisitor(targetJCheckConf, localHash);
```
I agree, fixed in latest version

edvbld commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckablePullRequest.java:
```diff
@@ -202,25 +204,30 @@ Hash amendManualReviewers(Hash commit, Namespace namespace, Hash original) throw
+        var additional = AdditionalConfiguration.get(original.get(), botUser, comments, reviewMerge);
+        return confOverride == null ?
+            JCheck.parseConfiguration(localRepo, hash, additional) :
+            JCheck.parseConfiguration(confOverride, additional);
```
Good idea, fixed in most recent change.

edvbld commented on jcheck/src/main/java/org/openjdk/skara/jcheck/JCheck.java:
```diff
@@ -295,4 +295,16 @@ public static Set<Check> checksFor(ReadOnlyRepository repository, Hash hash) thr
+                                conf,
+                                null);
+        return jcheck.checksForRange();
+    }
```
I think I would like to keep it for the tests since those tests will be exercising slightly different code paths (especially with the most recent version)

edvbld commented:
@erikj79 and @zhaosongzs I'm finally looping back to this and _think_ I addressed all or your comments and suggestions. Please have a look at this PR again, I think it is shaping up quite niceliy.

erikj79 commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:
```diff
@@ -1275,27 +1294,30 @@ private void checkStatus() {
+                // from the resulting commit. Not needed if we are overriding the JCheck configuration since
+                // then we won't use the one in the repo anyway.
+                if (workItem.bot.confOverrideRepository().isEmpty() &&
+                    (isFileUpdated(Path.of(".jcheck").resolve("conf"), localHash) || isJCheckConfUpdatedInMergePR)) {
```
Style nit. I think this way of creating a path looks a bit overly complex. I would suggest either just `Path.of(".jcheck/conf")` or split the elements like this `Path.of(".jcheck", "conf")`.

edvbld replied:
Sure, fixed in latest

erikj79 commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:
```diff
@@ -1219,30 +1218,49 @@ private void checkStatus() {
+
+                        var commitJCheckConf = checkablePullRequest.parseJCheckConfiguration(hash);
+                        var commitVisitor = checkablePullRequest.createVisitor(commitJCheckConf);
+                        if (isJCheckConfUpdatedInMergePR) {
```
Shouldn't this be conditional on the repo using an external jcheck conf?

erikj79 commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:
```diff
@@ -1275,27 +1294,30 @@ private void checkStatus() {
+                    jcheckType = "source jcheck";
+                    var localJCheckConf = checkablePullRequest.parseJCheckConfiguration(localHash);
+                    var localVisitor = checkablePullRequest.createVisitor(localJCheckConf);
+                    log.info("Run jcheck against localHash with JCHeck configuration from localHash");
```
```suggestion
                    log.info("Run JCheck against localHash with configuration from localHash");
```

edvbld replied:
Fixed in latest

erikj79 commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:
```diff
@@ -1275,27 +1294,30 @@ private void checkStatus() {
             }
 
-            var confFile = localRepo.lines(Path.of(".jcheck/conf"), localHash);
+            var confFile = localRepo.lines(Path.of(".jcheck").resolve("conf"), localHash);
```
Same style nit as above.

edvbld replied:
Fixed in latest

erikj79 replied:
This one wasn't changed. :)

edvbld replied:
Now it is 🙈

erikj79 reviewed (commented):
I like how this is shaping out. Just some minor issues left.

edvbld commented on bots/pr/src/main/java/org/openjdk/skara/bots/pr/CheckRun.java:
```diff
@@ -1219,30 +1218,49 @@ private void checkStatus() {
                                 .toList());
+
+                        var commitJCheckConf = checkablePullRequest.parseJCheckConfiguration(hash);
+                        var commitVisitor = checkablePullRequest.createVisitor(commitJCheckConf);
```
Agree, fixed in latest

edvbld commented:
Thanks @erikj79 for the thorough review! @zhaosongzs would you mind taking a look as well?

zhaosongzs commented:
> Thanks @erikj79 for the thorough review! @zhaosongzs would you mind taking a look as well?

I reviewed your latest patch yesterday and thought it pretty good. However, I hesitated to approve it because I thought ErikJ might discover some minor issues.

zhaosongzs reviewed (approved):
Looks very good!

edvbld commented:
Thanks @erikj79 and @zhaosongzs for reviewing!

/integrate
